src/lib/Imp.hs

```diff
@@ -454,544 +623,411 @@ splitDest (maybeDest, (Block decls ans)) = do
+  _ -> error $ "Not implemented " ++ pprint (dest, src)
+
+loadAnywhere :: IExpr -> ImpM IExpr
+loadAnywhere ptr = do
+  curDev <- asks curDevice
+  let (PtrType (_, addrSpace, ty)) = getIType ptr
+  case addrSpace of
+    Heap ptrDev | ptrDev /= curDev -> do
+      localPtr <- allocateStackSingleton ty
+      memcopy localPtr ptr (IIdxRepVal 1)
+      load localPtr
+    _ -> load ptr
+
+allocateStackSingleton :: IType -> ImpM IExpr
+allocateStackSingleton ty = allocateBuffer Stack False ty (IIdxRepVal 1)
 
 -- === Imp embedding ===
 
-embedBinOp :: MonadImp m instr => (Atom -> Atom -> Embed Atom) -> (IExpr -> IExpr -> m IExpr)
+embedBinOp :: (Atom -> Atom -> Embed Atom) -> (IExpr -> IExpr -> ImpM IExpr)
 embedBinOp f x y =
-  fromScalarAtom <$> fromEmbed (f (toScalarAtom (BaseTy bt) x) (toScalarAtom (BaseTy bt) y))
-  where (IValType bt) = impExprType x
+  fromScalarAtom <$> fromEmbed (f (toScalarAtom x) (toScalarAtom y))
 
-iaddI :: MonadImp m instr => IExpr -> IExpr -> m IExpr
+iaddI :: IExpr -> IExpr -> ImpM IExpr
 iaddI = embedBinOp iadd
 
-isubI :: MonadImp m instr => IExpr -> IExpr -> m IExpr
+isubI :: IExpr -> IExpr -> ImpM IExpr
 isubI = embedBinOp isub
 
-imulI :: MonadImp m instr => IExpr -> IExpr -> m IExpr
+imulI :: IExpr -> IExpr -> ImpM IExpr
 imulI = embedBinOp imul
 
-idivI :: MonadImp m instr => IExpr -> IExpr -> m IExpr
+idivI :: IExpr -> IExpr -> ImpM IExpr
 idivI = embedBinOp idiv
 
-impGet :: MonadImp m instr => IExpr -> IExpr -> m IExpr
-impGet ref i = case ref of
-  (IVar (_ :> IRefType t)) -> do
-    off <- t `offsetTo` i
-    let resTy = tabTypeGet t i
-    emitInstr $ liftInstr $ IOffset ref off resTy
-  _ -> error $ "impGet called with non-ref: " ++ show ref
+impOffset :: IExpr -> IExpr -> ImpM IExpr
+impOffset ref off = emitInstr $ IPrimOp $ PtrOffset ref off
 
-tabTypeGet :: ScalarTableType -> IExpr -> ScalarTableType
-tabTypeGet (TabTyAbs a) i = snd $ applyAbs a (toScalarAtom (absArgType a) i)
-tabTypeGet t _ = error $ "Not a non-scalar scalar table: " ++ pprint t
+cast :: IExpr -> BaseType -> ImpM IExpr
+cast x bt = emitInstr $ ICastOp bt x
 
-impOffset :: MonadImp m instr => IExpr -> IExpr -> ScalarTableType -> m IExpr
-impOffset ref off ty = case ref of
-  (IVar (_ :> IRefType _)) -> emitInstr $ liftInstr $ IOffset ref off ty
-  _ -> error $ "impOffset called with non-ref: " ++ show ref
+load :: IExpr -> ImpM IExpr
+load x = emitInstr $ IPrimOp $ PtrLoad x
 
-cast :: MonadImp m instr => IExpr -> BaseType -> m IExpr
-cast x bt = emitInstr $ liftInstr $ ICastOp (IValType bt) x
+memcopy :: IExpr -> IExpr -> IExpr -> ImpM ()
+memcopy dest src numel = emitStatement $ IInstr (Nothing , MemCopy dest src numel)
 
-load :: MonadImp m instr => IExpr -> m IExpr
-load x = emitInstr $ liftInstr $ Load x
+store :: HasCallStack => IExpr -> IExpr -> ImpM ()
+store dest src =
+  let (PtrType (_, _, ty)) = getIType dest
+  in if getIType src /= ty
+       then error "BAD!"
```

Comment:
I mean, this should be caught by some type checking later anyway, right?

Reply:
Oops, that's some leftover cruft from debugging.